### PR TITLE
LibJS: Flatten the bytecode representation and make interpreter fast

### PR DESCRIPTION
**Pull request by awesomekling**
tl;dr:
- Flatten the bytecode from disjoint basic blocks to contiguous byte sequence.
- Elide 0-length jumps
- Reorder codegen a bit to take advantage of 0-length jump elision
- Make bytecode interpreter threaded (as in "direct threaded", not multi-threaded)
- Make instructions smaller in various ways

Result: everything much faster!

Benchmark results before/after:
```
Suite       Test                                   Speedup  Old (Mean ± Range)           New (Mean ± Range)
----------  -----------------------------------  ---------  ---------------------------  ---------------------------
Kraken      ai-astar.js                              1.707  4.960 ± 4.950 … 4.970        2.905 ± 2.860 … 2.950
Kraken      audio-beat-detection.js                  1.544  2.610 ± 2.560 … 2.660        1.690 ± 1.690 … 1.690
Kraken      audio-dft.js                             1.441  2.335 ± 2.300 … 2.370        1.620 ± 1.620 … 1.620
Kraken      audio-fft.js                             1.502  2.350 ± 2.330 … 2.370        1.565 ± 1.540 … 1.590
Kraken      audio-oscillator.js                      1.409  2.565 ± 2.550 … 2.580        1.820 ± 1.810 … 1.830
Kraken      imaging-darkroom.js                      1.327  3.325 ± 3.290 … 3.360        2.505 ± 2.500 … 2.510
Kraken      imaging-desaturate.js                    1.31   3.525 ± 3.520 … 3.530        2.690 ± 2.670 … 2.710
Kraken      imaging-gaussian-blur.js                 1.616  16.785 ± 16.710 … 16.860     10.385 ± 10.360 … 10.410
Kraken      json-parse-financial.js                  1      0.130 ± 0.130 … 0.130        0.130 ± 0.130 … 0.130
Kraken      json-stringify-tinderbox.js              1      0.170 ± 0.170 … 0.170        0.170 ± 0.170 … 0.170
Kraken      stanford-crypto-aes.js                   1.378  1.130 ± 1.120 … 1.140        0.820 ± 0.820 … 0.820
Kraken      stanford-crypto-ccm.js                   1.291  0.910 ± 0.900 … 0.920        0.705 ± 0.700 … 0.710
Kraken      stanford-crypto-pbkdf2.js                1.502  1.915 ± 1.900 … 1.930        1.275 ± 1.270 … 1.280
Kraken      stanford-crypto-sha256-iterative.js      1.385  0.720 ± 0.720 … 0.720        0.520 ± 0.520 … 0.520
Octane      box2d.js                                 1.276  4.320 ± 4.320 … 4.320        3.385 ± 3.370 … 3.400
Octane      code-load.js                             1.007  2.150 ± 2.140 … 2.160        2.135 ± 2.130 … 2.140
Octane      crypto.js                                1.368  10.120 ± 9.930 … 10.310      7.395 ± 7.360 … 7.430
Octane      deltablue.js                             1.494  3.040 ± 3.040 … 3.040        2.035 ± 2.030 … 2.040
Octane      earley-boyer.js                          1.069  22.055 ± 21.950 … 22.160     20.635 ± 20.570 … 20.700
Octane      gbemu.js                                 1.109  5.120 ± 5.120 … 5.120        4.615 ± 4.610 … 4.620
Octane      mandreel.js                              1.423  24.795 ± 24.600 … 24.990     17.425 ± 17.190 … 17.660
Octane      navier-stokes.js                         1.368  4.445 ± 4.440 … 4.450        3.250 ± 3.230 … 3.270
Octane      pdfjs.js                                 1.133  3.845 ± 3.840 … 3.850        3.395 ± 3.380 … 3.410
Octane      raytrace.js                              1.028  8.215 ± 8.140 … 8.290        7.995 ± 7.990 … 8.000
Octane      regexp.js                                1.035  24.300 ± 24.090 … 24.510     23.470 ± 23.200 … 23.740
Octane      richards.js                              1.002  2.025 ± 2.020 … 2.030        2.020 ± 2.020 … 2.020
Octane      splay.js                                 1.01   2.900 ± 2.900 … 2.900        2.870 ± 2.860 … 2.880
Octane      typescript.js                            1.131  52.625 ± 52.570 … 52.680     46.535 ± 46.470 … 46.600
Octane      zlib.js                                  1.74   198.025 ± 196.870 … 199.180  113.820 ± 112.970 … 114.670
Kraken      Total                                    1.508  43.430                       28.800
Octane      Total                                    1.41   367.980                      260.980
All Suites  Total                                    1.42   411.410                      289.780
```

### CODE REVIEW

File: Userland/Libraries/LibJS/Bytecode/Interpreter.cpp [hunk 1]
```diff
@@ -330,7 +330,14 @@ Interpreter::HandleExceptionResponse Interpreter::handle_exception(size_t& progr
     VERIFY_NOT_REACHED();
 }
 
-void Interpreter::run_bytecode(size_t entry_point)
+// FIXME: GCC takes a *long* time to compile with flattening, and it will time out our CI. :|
+#if defined(AK_COMPILER_CLANG)
```
Comment (trflynn89) [inline Userland/Libraries/LibJS/Bytecode/Interpreter.cpp:334]: Is GCC slow in general, or just with sanitizers enabled? If (hopefully) the latter, we could just detect GCC+ASAN here

**Review by trflynn89 — APPROVED** [review]

File: Userland/Libraries/LibJS/Bytecode/Interpreter.cpp (hunk 1 above)
Comment (awesomekling) [reply Userland/Libraries/LibJS/Bytecode/Interpreter.cpp:334]: Slow in general, it takes several minutes to build Interpreter.cpp on my machine, so I am reasonably sure it will kill CI :(

File: Userland/Libraries/LibJS/Bytecode/Generator.cpp [hunk 2]
```diff
@@ -108,6 +110,11 @@ CodeGenerationErrorOr<NonnullGCPtr<Executable>> Generator::generate(VM& vm, ASTN
         }
 
         block_offsets.set(block.ptr(), bytecode.size());
+
+        for (auto& source_map_entry : block->source_map()) {
+            source_map.set(bytecode.size() + source_map_entry.key, source_map_entry.value);
+        }
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibJS/Bytecode/Generator.cpp:116]:
```suggestion
        for (auto [offset, source_record] : block->source_map())
            source_map.set(bytecode.size() + offset, source_record);
```

File: Userland/Libraries/LibJS/Bytecode/Interpreter.cpp [hunk 3]
```diff
@@ -340,7 +340,7 @@ void Interpreter::run_bytecode(size_t entry_point)
     auto* locals = running_execution_context.locals.data();
     auto& accumulator = this->accumulator();
     auto& executable = current_executable();
-    auto& bytecode = executable.bytecode;
+    auto const* bytecode = executable.bytecode.data();
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibJS/Bytecode/Interpreter.cpp:346]: This change seems unrelated
(LibJS/Bytecode: Keep instruction source mappings in Executable)

**Review by Hendiadyoin1 — COMMENTED** [review]
Nit/opinion: I'd like it to retain the ability to dump the basic blocks, as some reasoning about code might be simpler there, as well as iff we bring back an optimization pipeline it would be nice to be able to work on basic blocks in some passes, while others may handle the flattened version fine
I see a GFC based reordering pass being quite useful for example

Anyway future stuff that can probably be shoved in later without much hassle

Great work
